tools/base/fpgainfo/board.c

```diff
@@ -0,0 +1,413 @@
+
+		if (platform_data_table[i].dl_handle) {
+
+			res = dlclose(platform_data_table[i].dl_handle);
+			if (res) {
+				char *err = dlerror();
+				OPAE_ERR("dlclose failed with %d %s", res, err ? err : "");
+			}
+		} //end if
+
+	} // end for
+
+	return res;
+}
+
+/*
+ * Print help
+ */
+void mac_help(void)
+{
+	printf("\nPrint MAC information\n"
+		"        fpgainfo mac [-h]\n"
+		"                -h,--help           Print this help\n"
+		"\n");
+}
+
+#define MAC_GETOPT_STRING ":h"
+int parse_mac_args(int argc, char *argv[])
+{
+	struct option longopts[] = {
+		{"help", no_argument, NULL, 'h'},
+		{0, 0, 0, 0},
+	};
+	int getopt_ret;
+	int option_index;
+
+	optind = 0;
+	while (-1 != (getopt_ret = getopt_long(argc, argv, MAC_GETOPT_STRING,
+		longopts, &option_index))) {
+		const char *tmp_optarg = optarg;
+
+		if (optarg && ('=' == *tmp_optarg)) {
+			++tmp_optarg;
+		}
+
+		switch (getopt_ret) {
+		case 'h':   /* help */
+			mac_help();
+			return -1;
+
+		case ':':   /* missing option argument */
+			fprintf(stderr, "Missing option argument\n");
+			mac_help();
+			return -1;
+
+		case '?':
+		default:    /* invalid option */
+			fprintf(stderr, "Invalid cmdline options\n");
+			mac_help();
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+fpga_result mac_filter(fpga_properties *filter, int argc, char *argv[])
+{
+	fpga_result res = FPGA_INVALID_PARAM;
+
+	if (0 == parse_mac_args(argc, argv)) {
+		res = fpgaPropertiesSetObjectType(*filter, FPGA_DEVICE);
+		fpgainfo_print_err("Setting type to FPGA_DEVICE", res);
+	}
+	return res;
+}
+
+fpga_result mac_command(fpga_token *tokens, int num_tokens, int argc,
+	char *argv[])
+{
+	(void)argc;
+	(void)argv;
+	fpga_result res = FPGA_OK;
+	fpga_properties props;
+
+	int i = 0;
+	for (i = 0; i < num_tokens; ++i) {
+		res = fpgaGetProperties(tokens[i], &props);
+		ON_FPGAINFO_ERR_GOTO(res, out_destroy, "Reading properties from token");
+
+		//print_mac_info(props);
+		mac_info(tokens[i]);
+		fpgaDestroyProperties(&props);
+	}
+
+	return res;
+
+out_destroy:
+	fpgaDestroyProperties(&props);
+	return res;
+}
+
+
+//phy
+
+/*
+ * Print help
+ */
+void phy_help(void)
+{
+	printf("\nPrint PHY information\n"
+		"        fpgainfo phy [-h] [-G <group-number>]\n"
+		"                -h,--help           Print this help\n"
+		"                -G,--group          Select PHY group {0,1,all}\n"
+		"\n");
+}
+
+#define PHY_GETOPT_STRING ":G:h"
+int group_num;
+int parse_phy_args(int argc, char *argv[])
+{
+	struct option longopts[] = {
+		{"group", required_argument, NULL, 'G'},
+		{"help", no_argument, NULL, 'h'},
+		{0, 0, 0, 0},
+	};
+	int getopt_ret;
+	int option_index;
+
+	/* default configuration */
+	group_num = -1;
+
+	optind = 0;
+	while (-1 != (getopt_ret = getopt_long(argc, argv, PHY_GETOPT_STRING,
+		longopts, &option_index))) {
+		const char *tmp_optarg = optarg;
+
+		if (optarg && ('=' == *tmp_optarg)) {
+			++tmp_optarg;
+		}
+
+		switch (getopt_ret) {
+		case 'G':
+			if (NULL == tmp_optarg)
+				break;

```

Comment:
If we break here, then the fn will return 0.  Is that what we want?

Reply:
fixed